service/history/shard/coordinator.go

```diff
@@ -0,0 +1,214 @@
+		shutdownChan chan struct{}
+
+		config *config.Config
+		logger log.Logger
+	}
+
+	request struct {
+		shardID int
+		markers []*replicator.FailoverMarkerAttributes
+		respCh  chan error
+	}
+
+	failoverRecord struct {
+		failoverVersion int64
+		shards          map[int]struct{}
+	}
+)
+
+// NewCoordinator initialize a failover coordinator
+func NewCoordinator(
+	config *config.Config,
+	logger log.Logger,
+) Coordinator {
+	return &coordinatorImpl{
+		status:       common.DaemonStatusInitialized,
+		recorder:     make(map[string]*failoverRecord),
+		sendChan:     make(chan *request, sendChanBufferSize),
+		receiveChan:  make(chan *request, receiveChanBufferSize),
+		shutdownChan: make(chan struct{}),
+		config:       config,
+		logger:       logger,
+	}
+}
+
+func (c *coordinatorImpl) Start() {
+
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	go c.receiveFailoverMarkersLoop()
+	go c.sendFailoverMarkerLoop()
+}
+
+func (c *coordinatorImpl) Stop() {
+
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(c.shutdownChan)
+}
+
+func (c *coordinatorImpl) SendFailoverMarkers(
+	shardID int,
+	markers []*replicator.FailoverMarkerAttributes,
+) <-chan error {
+
+	respCh := make(chan error, 1)
+	c.sendChan <- &request{
+		shardID: shardID,
+		markers: markers,
+		respCh:  respCh,
+	}
+
+	return respCh
+}
+
+func (c *coordinatorImpl) ReceiveFailoverMarkers(
+	shardID int,
+	markers []*replicator.FailoverMarkerAttributes,
+) {
+
+	c.receiveChan <- &request{
+		shardID: shardID,
+		markers: markers,
+	}
+}
+
+func (c *coordinatorImpl) receiveFailoverMarkersLoop() {
+
+	for {
+		select {
+		case <-c.shutdownChan:
+			return
+		case request := <-c.receiveChan:
+			for _, marker := range request.markers {
+				domainID := marker.GetDomainID()
+
+				if record, ok := c.recorder[domainID]; ok {
+					// if the local failover version is smaller than the new received marker,
+					// it means there is another failover happened and the local one should be invalid.
+					if record.failoverVersion < marker.GetFailoverVersion() {
+						delete(c.recorder, domainID)
+					}
+				}
+
+				if _, ok := c.recorder[domainID]; !ok {
+					// initialize the failover record
+					c.recorder[marker.GetDomainID()] = &failoverRecord{
+						failoverVersion: marker.GetFailoverVersion(),
+						shards:          make(map[int]struct{}),
+					}
+					return
+				}
+
+				record := c.recorder[domainID]
+				record.shards[request.shardID] = struct{}{}
+				if len(record.shards) == c.config.NumberOfShards {
+					// TODO: update domain from pend_active to active
+					delete(c.recorder, domainID)
+				}
+			}
+		}
+	}
+}
+
+func (c *coordinatorImpl) sendFailoverMarkerLoop() {
+
+	timer := time.NewTimer(backoff.JitDuration(
+		c.config.FailoverMarkerHeartbeatInterval(),
+		c.config.FailoverMarkerHeartbeatTimerJitterCoefficient(),
+	))
```

Comment:
what is the point of grouping the failover?

Reply:
rename the function to heartbeatFailoverMarker. This is when each shard receives the failover marker, it will heartbeat the markers to the coordinator. So here grouping the heartbeat to reduce RPC calls. Each host will aggregate the heartbeat and send them to the coordinator.

Reply:
The heartbeat is to track the liveness of each shard?

Reply:
No. Each shard heartbeats last X failover markers to the coordinator. The markers persist at each shard. The coordinator does not persist anything (to DB).